### ISSUE TITLE
Istio discovery failing to form base mesh in GKE

### ISSUE THREAD

**Issue by gabClark** [istio/istio#39673]
### Bug Description

We have istio cluster using 1.9.8 and pushing to upgrade out of end of life.
Istio-operator is used to deploy it. Cluster has been operational historically in previous versions of istio but having this issue with version 1.10 and higher.
After upgrading and the istiod and gateways changing to 1.10.6 some of our deployments went into unhealthy state. upon closer inspection using `istioctl proxy-config routes` and `istioctl proxy-config endpoint` revealed that the proxy configuration for said pods was missing dozens and dozen of our regular entries from the cluster in that same namespace. things like other apis, elastic-search, etc. Upon downgrading back to 1.9.8 routes and endpoints were immediately populated again.
It looks like as discovery is going over the "main" namespace that has most of our components something causes it to fail and EDS, RDS and company do not push full correct  configuration. 
Here is a sample snapshot of pilot pushes during a window of 1.10.6
![image](https://user-images.githubusercontent.com/93932584/176212883-192ba552-6d6d-4ecf-aa6f-e56f3904c920.png)

Pushes area fraction of what they should be, then as soon as downgrade kicks at 10:44 it is restored. However no errors are reported, even looking at istiod logs directly there are many `info ads LDS: PUSH request for x` or the other discovery services but no warnings or errors at all. 
Additionally `istioctl proxy-status` would report CDS, LDS, EDS, and RDS are all `SYNCED` so it is my understanding istio even thinks it did everything it had to.

### Version

```prose
istioctl version
client version: 1.10.6
control plane version: 1.9.8
data plane version: 1.9.8 (77 proxies)
---
kubectl version --short
Client Version: v1.21.5
Server Version: v1.21.11-gke.1100
```

### Additional Information

```
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-control-plane
  namespace: istio-system
spec:
  profile: default
  components:
    egressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 2
          minReplicas: 2
        resources:
          limits:
            cpu: "2"
            memory: 2Gi
          requests:
            cpu: 350m
            memory: 771Mi
        service:
          ports:
           ....
      name: istio-egressgateway
      namespace: istio-system
    ingressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 2
          minReplicas: 2
        resources:
          limits:
            cpu: "2"
            memory: 1Gi
          requests:
            cpu: 350m
            memory: 384Mi
        service:
          loadBalancerIP: our_ip
      name: istio-ingressgateway
      namespace: istio-system
    pilot:
      enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 5
          minReplicas: 2
        env:
          - name: PILOT_ENABLE_INBOUND_PASSTHROUGH # tried this from the 1.10  changes docs
            value: "false"

  hub: gcr.io/istio-release
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
    accessLogFormat: '{"authority":"%REQ(:AUTHORITY)%","grpc_status":"%GRPC_STATUS%","hostname":"%HOSTNAME%","message":"%RESPONSE_CODE%
      %REQ(:METHOD)% %REQ(:AUTHORITY)%%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%","method":"%REQ(:METHOD)%","path":"%REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%","protocol":"%PROTOCOL%","requested_server_name":"%REQUESTED_SERVER_NAME%","response_code":"%RESPONSE_CODE%","response_flags":"%RESPONSE_FLAGS%","severity":"INFO","start_time":"%START_TIME%","upstream_host":"%UPSTREAM_HOST%","upstream_local_address":"%UPSTREAM_LOCAL_ADDRESS%","user_agent":"%REQ(USER-AGENT)%"}'
  # Tried these too.
  values:
     gateways:
       istio-egressgateway:
         runAsRoot: true
     sidecarInjectorWebhook:
       useLegacySelectors: true
```

**Comment by gabClark** [comment]
Also tried now on latest istio(1.14.1) and experience the same behavior. It is likely that this is some incompatibility with out setup or environment but it is kind of obscure and hard to determine what could it be.

**Comment by litong01** [comment]
@gabClark when the errors happened, did you see that your pod still having the sidecar or sidecars were missing as well? We've found that during upgrades, when pods also get restarted, sidecars will disappear because the sidecar injection webhook might have been disabled momentarily, that causes sidecar injection to fail, thus endpoint goes away.

**Comment by gabClark** [comment]
> @gabClark when the errors happened, did you see that your pod still having the sidecar or sidecars were missing as well? We've found that during upgrades, when pods also get restarted, sidecars will disappear because the sidecar injection webhook might have been disabled momentarily, that causes sidecar injection to fail, thus endpoint goes away.

Hello and thanks, sidecars were perfectly fine and that was part of the confusion. No signs of something wrong at all, and indeed it turns out nothing was broken. 

Recently I took a deeper dive and parsed deeper into the root of it and we identified(and now tested) that the issue originated from 1.10.3 bugfix of wildcards for egress. Seems like we were depending on this "feature" and the change in behavior made it so our apis(all of which had sidecars configurations) were no longer discovering other things in cluster properly.
